File: backend/src/domain/game-definition.ts
import {
  DisplayText,
  GameType,
  Metadata,
  ScorecardDefinitionId
} from "./common";

export type GameEventCategory =
  | "score"
  | "attempt"
  | "penalty"
  | "phase"
  | "system"
  | "custom";

export interface GameEventType extends DisplayText {
  id: string;
  category: GameEventCategory;
  affectsScore?: boolean;
  countsAsAttempt?: boolean;
  statKey?: string;
  metadata?: Metadata;
}

export interface GamePhaseDefinition extends DisplayText {
  id: string;
  sequence: number;
  isOvertime?: boolean;
  metadata?: Metadata;
}

export type StatScope = "player" | "team" | "match" | "tournament";
export type StatValueType = "count" | "number" | "percentage" | "duration";

export interface StatDefinition extends DisplayText {
  key: string;
  scope: StatScope;
  valueType: StatValueType;
  precision?: number;
  metadata?: Metadata;
}

export interface GameDefinition {
  gameType: GameType;
  displayName: string;
  phases: GamePhaseDefinition[];
  eventTypes: GameEventType[];
  stats: StatDefinition[];
  scorecardDefinitionId: ScorecardDefinitionId;
  metadata?: Metadata;
}

File: Ruski Report/Models/GameDefinition.swift
//
//  GameDefinition.swift
//  Ruski Report
//

import Foundation

nonisolated struct GameDefinition: Identifiable, Equatable {
    var id: String { gameType }

    let gameType: String
    let displayName: String
    let scorecardDefinitionId: String
    let phases: [GamePhaseDefinition]
    let eventTypes: [GameEventTypeDefinition]
    let stats: [GameStatDefinition]
}

nonisolated struct GamePhaseDefinition: Identifiable, Equatable {
    let id: String
    let label: String
    let sequence: Int
}

nonisolated struct GameEventTypeDefinition: Identifiable, Equatable {
    let id: String
    let label: String
    let category: String
    let affectsScore: Bool
    let countsAsAttempt: Bool
    let statKey: String?
}

nonisolated struct GameStatDefinition: Identifiable, Equatable {
    var id: String { key }

    let key: String
    let label: String
    let scope: String
    let valueType: String
}
